cmd/pki.go

```diff
@@ -0,0 +1,354 @@
+// Copyright © 2018 NAME HERE <EMAIL ADDRESS>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"github.com/spf13/cobra"
+	"log"
+	"math/big"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/netm4ul/netm4ul/core/config"
+	"github.com/spf13/viper"
+	"path/filepath"
+	"strconv"
+)
+
+var (
+	nNodes               uint
+	serverID             string
+	ecdsaCurve           string
+	keepPrivCA           bool
+	certDuration         time.Duration
+	organisationSubject  string
+	certificateDirectory string
+)
+
+// pkiCmd represents the pki command
+var pkiCmd = &cobra.Command{
+	Use:   "pki",
+	Short: "Set up PKI (CA, Server and clients)",
+	Long: `This command builds a Public Key Infrastructure (PKI) for mutual TLS authentication between your server and the nodes.
+	A Certificate Authority (CA) is created, signing the server (master node) certificate, as well as all the client nodes at first deploy time.
+	The CA certificate is shipped to server and clients along with own certificate and private key of an entity, thus enabling mutual authentication on TLS handshake.
+	
+	The parameters are to be specified in the configuration file or inline.
+	
+	The master node has probably an IP address as hostname, if you run your instance from a box without DNS indexing.
+	
+	The client nodes don't need a publicly known hostname, but you may assign them a name that suits your use (e.g. your logs may see "client_x successfully connected to API"")`,
+
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		config.LoadConfig(configPath)
+
+		if serverID == "" {
+			serverID = config.Config.Server.IP
+		}
+
+	},
+
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("Setting up PKI at " + certificateDirectory)
+		pkiSetup(organisationSubject, nNodes, certDuration, ecdsaCurve, certificateDirectory, serverID)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(pkiCmd)
```

Comment:
I think we should not add this command to the `rootCmd` but on the `setupCmd`.
This is part of the setup phase.

Reply:
ok. I wasn't sure. Will do.